Data further refactor

dirvine opened the pull request:
wip - this PR is on top of the data_refactor PR

dirvine committed:
feat: add Pointer type and make LinkedList outputs optional

* Add new Pointer type for network addressing
* Add PointerAddress type and related network address handling
* Implement signature verification for Pointer records
* Add validation and storage handling for Pointer records in ant-node
* Add log markers for Pointer operations
* Make LinkedList outputs field optional by wrapping in Option<Vec>
* Update LinkedList constructors and signature calculation to handle optional outputs
* Fix transaction tests to use .into() for empty output vectors
* Update ant-networking to handle split records for both Pointer and optional outputs
* Maintain signature verification and count-based selection for split records
* Add InvalidSignature error variant to handle Pointer validation failures

This change introduces a new Pointer type for network addressing while also making
LinkedList transactions more flexible by allowing them to be created without outputs.
The networking layer has been updated to handle both changes, including proper
split record handling, signature verification, and storage validation. All changes
maintain backward compatibility with existing functionality.

dirvine committed:
feat: add Python bindings for ChunkAddress and Pointer types, update version numbers, and add tests

dirvine committed:
refactor: update self_encryption to use streaming API - Replace DataMapLevel with direct streaming API usage - Update client utils to use streaming_decrypt_from_storage - Add detailed implementation and schedule documentation - Clean up unused imports and improve error handling - Update archive.rs to work with new self_encryption API

dirvine committed:
docs: add architecture documentation and update cargo.toml

dirvine committed:
refactor: remove WASM support - Remove WASM dependencies and crate-type - Remove WASM test file and documentation - Update README to remove WASM references

dirvine committed:
fix: improve local network testing - Require only one peer when in local mode instead of CLOSE_GROUP_SIZE - Remove global address check for local testing - Fix test configuration to properly use local feature - Clean up unused imports and variables

dirvine committed:
refactor(network): Simplify LocalNode implementation

- Refactor LocalNode to use tokio async/await consistently

- Add proper timeouts and cleanup in start/stop methods

- Improve process handling with kill_on_drop

- Remove unused peer discovery and event handling

- Clean up imports and remove warnings

- Add comprehensive test with proper cleanup

dirvine committed:
feat(network): update LocalNode to use antnode binary with mDNS discovery

- Replace ipfs with antnode binary for node operations

- Implement proper mDNS peer discovery

- Add proper error handling and logging

- Increase test timeouts for node startup

- Fix compilation issues and dependencies

dirvine committed:
refactor: make local mode runtime configurable

- Remove local feature flag dependency, making it a runtime configuration

- Ensure bootstrap cache is not written to in local mode

- Fix builder pattern usage in NetworkBuilder

- Add proper config handling in deprecated connect method

- Update tests to work with runtime local mode configuration

This change makes the local mode behavior more explicit and configurable at runtime, rather than being a compile-time feature flag. It also ensures that when in local mode, we don't pollute the bootstrap cache with local-only peers.

dirvine committed:
refactor: improve network interface detection and QUIC support

- Update find_local_ip to use UDP sockets for QUIC protocol\n- Add comprehensive IP validation (non-loopback, private range, etc.)\n- Update multiaddr format to use QUIC instead of TCP\n- Fix error handling in network code and tests\n- Add test coverage for IP validation and QUIC multiaddr format\n\nThis change ensures we use the correct network interface for QUIC connections and validates the IP address meets all requirements for P2P networking.

dirvine committed:
refactor: Update local network to use find_local_ip() and QUIC

dirvine committed:
refactor: update NetworkBuilder usage to handle local mode and bootstrap cache correctly

dirvine committed:
fix(tests): improve local node test reliability and debugging

- Fix command-line argument order for antnode, placing global flags before subcommand
- Add stdout/stderr capture to LocalNode for better debugging
- Split node startup into first and secondary node functions
- Add node status checking to detect early failures
- Reduce test timeout from 120s to 30s for faster feedback
- Add better logging of node addresses and connection status

The test now properly starts two nodes in a local network configuration,
with the first node having the --first flag and the second node joining
as a peer. Output capturing helps diagnose startup and connection issues.

dirvine committed:
refactor: improve error handling and test infrastructure

This commit enhances error handling and test infrastructure across the codebase:

Error Handling:

- Add Serialization variant to CostError for consistent error handling

- Fix error conversions for serialization errors in archive handling

- Ensure proper error propagation in client operations

Module Organization:

- Fix Client import in pointer.rs to use super::Client

- Re-export PutError in lib.rs for test accessibility

- Clean up import paths for better code organization

Method Improvements:

- Make check_write_access public for testing purposes

- Fix wallet() method access to use proper method call

Test Infrastructure:

- Add local EVM network configuration for testing

- Create EVM testnet data CSV file with test configuration

- Fix test environment setup for client mode tests

This change improves code maintainability and test reliability while ensuring consistent error handling across the codebase.

dirvine committed:
refactor(test-utils): use temporary directory for EVM wallet tests

- Add tempfile dependency for temporary directory management

- Create setup_test_environment function to handle test environment setup

- Use temporary directory for EVM testnet data and configuration

- Clean up imports and improve code organization

dirvine committed:
refactor: move find_local_ip to ant-bootstrap utils

- Added utils module to ant-bootstrap with find_local_ip function

- Removed ant-networking dependency from ant-bootstrap to avoid circular dependency

- Updated address_format_tests to use the new utils function

- Cleaned up unused imports

dirvine committed:
feat: add local test suite and CI workflow

- Add test-local.sh script for running local tests

- Add GitHub Actions workflow for running local tests

- Add linked_list tests

- Update client implementation for linked list operations

- Clean up unused imports and fix linter warnings

dirvine committed:
fix: update test scripts to handle port conflicts

- Add cleanup before starting EVM testnet

- Set specific RPC port for EVM testnet

dirvine committed:
feat: improve test scripts to handle existing EVM network

- Add port checks before starting EVM testnet

- Use consistent port 4343 across scripts

- Only clean up antnode processes, preserve EVM network

dirvine committed:
refactor: remove transaction.rs test file

- Functionality moved to linked_list.rs

dirvine committed:
feat: expose LinkedList types in public API

- Add LinkedList and LinkedListAddress to public exports

dirvine committed:
refactor: update transaction references to linked list - Replace transaction terminology with linked list in error messages and comments - Update Python bindings to support linked list functionality - Add LinkedList and LinkedListAddress Python classes - Update Python documentation with linked list API reference - Remove deprecated transaction references in registers.rs

dirvine committed:
feat: add nodejs typescript bindings structure - Add initial TypeScript project setup - Create basic client interface with linked list support - Add comprehensive type definitions - Set up Jest testing framework - Add documentation and examples

dirvine committed:
docs: add detailed Node.js bindings documentation - Add installation and setup instructions - Include TypeScript configuration details - Add API reference and examples - Document best practices and testing

dirvine committed:
docs: add comprehensive API documentation and guides for Node.js, Python, and Rust, including local network setup, EVM integration, and testing guides

dirvine committed:
docs: set up MkDocs with Material theme and GitHub Actions deployment

dirvine committed:
docs: update README and add comprehensive contributing guide with documentation instructions

dirvine committed:
docs: add comprehensive data storage guide covering self-encryption and scratchpad features

dirvine committed:
docs: add Files, Directories, Archive, and Vault documentation to API references

- Added File and Directory management APIs

- Added Archive creation and extraction APIs

- Added secure Vault storage APIs

- Added comprehensive type definitions and examples

- Maintained language-specific idioms for Node.js, Python, and Rust

dirvine committed:
chore: remove node_modules from git tracking and add to gitignore

dirvine committed:
chore: add node_modules to gitignore

dirvine committed:
refactor: move local feature to test feature to ensure production builds don't include local mode

dirvine committed:
fix: remove node_modules

dirvine committed:
fix: update test scripts to use local feature for ant-node and test feature for autonomi

dirvine committed:
fix: update test files to use test feature and fix imports

dirvine committed:
refactor: improve test-local.sh script with configuration and helper functions

dirvine committed:
feat: add start-local-network.sh script and documentation for local development

dirvine committed:
docs: add comprehensive payments guide for put operations

dirvine committed:
docs: add local development guide to navigation

dirvine committed:
docs: remove incorrect Docker requirement and improve local network guide

dirvine committed:
docs: enhance EVM network and monitoring/debugging sections in local network guide

dirvine committed:
doc: exclude site packages

dirvine committed:
docs: add client modes guide and update API documentation

dirvine committed:
docs: update mkdocs navigation and fix broken links

dirvine committed:
docs: add comprehensive data types guide and reorganize documentation structure

dirvine committed:
docs: enhance data types guide with detailed technical information

dirvine committed:
docs: update documentation styling to match autonomi.com website design

dirvine committed:
docs: add library documentation and update API references

dirvine committed:
chore: update mkdocs emoji configuration

dirvine committed:
chore: simplify mkdocs configuration

dirvine committed:
fix: ignore cache files

dirvine committed:
docs: add comprehensive development guides for web, quantum security, error handling, data science, and Rust performance

dirvine committed:
docs: fix code block styling and broken documentation links

dirvine committed:
docs: update quantum security guide to clarify BLS vs self-encryption security properties, remove pointer design doc

dirvine committed:
docs: update API documentation styling and content

dirvine committed:
docs: reorganize API reference to show all languages together

dirvine committed:
docs: fix relative links in API documentation

dirvine committed:
docs: update API documentation with improved descriptions and light mode code blocks

dirvine committed:
docs: consolidate API documentation into a single file with language selector

dirvine committed:
docs: consolidate API documentation into single file with all language examples

dirvine committed:
docs: update mkdocs navigation for consolidated API documentation

dirvine committed:
docs: remove unused documentation files and reorganize structure

dirvine committed:
docs: update API reference links to point to consolidated documentation

dirvine committed:
docs: update API documentation to use tabbed code blocks for language selection

dirvine committed:
docs: update installation and API docs to reflect current implementation

dirvine committed:
docs: update navigation to include all guides and improve structure

dirvine committed:
docs: add build from source instructions for all languages

dirvine committed:
docs: add comprehensive API documentation for all components

dirvine committed:
docs: fix broken links in documentation

dirvine committed:
doc: update nav

dirvine committed:
docs: reorganize documentation structure and fix links

dirvine committed:
fix: moved docs

dirvine committed:
doc: api nav

dirvine committed:
docs: update self-encryption API docs with streaming operations and chunk store implementations

dirvine committed:
docs: remove Quick Start guide and references

dirvine committed:
docs: add core concepts to Quick Links section

dirvine committed:
docs: update main documentation index with improved installation and usage instructions

grumbach commented on ant-bootstrap/tests/cli_integration_tests.rs:
```diff
@@ -176,3 +182,45 @@ async fn test_test_network_peers() -> Result<(), Box<dyn std::error::Error>> {
+
+    // ... rest of the test ...
+    Ok(())
+}
```
Agreed that this would be a nice feature! Maybe add a TODO comment in the tests here so we don't forget them as is?

grumbach commented on ant-evm/src/amount.rs:
```diff
@@ -67,7 +67,7 @@ impl AttoTokens {
 
     /// Converts the Nanos into bytes
     pub fn to_bytes(&self) -> Vec<u8> {
```
Here the comment still refers to Nanos

dirvine replied:
Yea I missed that one. This file was failing tests mind you, that's the only reason I ended up making changes to it.

grumbach commented on ant-evm/src/amount.rs:
```diff
@@ -67,7 +67,7 @@ impl AttoTokens {
     /// Converts the Nanos into bytes
     pub fn to_bytes(&self) -> Vec<u8> {
-        self.0.as_le_bytes().to_vec()
+        self.0.to_be_bytes::<32>().to_vec()
```
Any specific reasons to move to big endian and 32 here?

dirvine replied:
Really network type endianness (most significant bit first) The 32 states we want exactly 32bytes (256 bits)
Otherwise we risk returning a variable length vec, well we would if using different types at least. So here we force the 32bytes. 

In short we guarantee a 32byte network endian number here really.

grumbach commented on ant-networking/Cargo.toml:
```diff
@@ -58,7 +59,6 @@ prometheus-client = { version = "0.22", optional = true }
 rand = { version = "~0.8.5", features = ["small_rng"] }
 rayon = "1.8.0"
 rmp-serde = "1.1.1"
-self_encryption = "~0.30.0"
```
Glad we were able to trim dependencies here!

grumbach commented on ant-networking/src/driver.rs:
```diff
@@ -300,6 +300,11 @@ impl NetworkBuilder {
         }
     }
 
+    pub fn local(mut self, local: bool) -> Self {
+        self.local = local;
+        self
+    }
+
```
This is a GREAT pattern!

dirvine committed:
doc: docstring nano -> atto

dirvine committed:
doc: API ref

dirvine committed:
doc: index layout

grumbach commented on autonomi/README_WASM.md:
```diff
@@ -1,95 +0,0 @@
-# Autonomi JS API
-
-Note: the JS API is experimental and will be subject to change.
```
Are we sure we want to delete these WASM docs now?

dirvine replied:
yes we gave up on wasm for now. I hope we introduce it at a later stage, but we need to ensure we do not hamper network or disk access when we do

dirvine committed:
chore: update Cargo.lock after rebase

grumbach commented on autonomi/src/client/data/public.rs:
```diff
@@ -44,100 +49,72 @@ impl Client {
-            };
-            if let Err(err) = channel.send(ClientEvent::UploadComplete(summary)).await {
-                error!("Failed to send client event: {err:?}");
+        match &self.mode {
+            ClientMode::ReadWrite(_) => {
+                let now = ant_networking::target_arch::Instant::now();
```
Why introduce a mode here?

dirvine replied:
The idea is we have a read only client (no wallet etc.) and a full client that can put so needs a wallet

grumbach commented on autonomi/src/client/error.rs:
```diff
@@ -0,0 +1,123 @@
+use thiserror::Error;
+use xor_name::XorName;
+
+/// Errors that can occur during data storage operations
+#[derive(Debug, thiserror::Error)]
+pub enum PutError {
+    /// No wallet available for write operations
+    #[error("Write operations require a wallet. Use upgrade_to_read_write to add a wallet.")]
+    NoWallet,
+    /// Network-related error
+    #[error("Network error: {0}")]
+    Network(#[from] NetworkError),
```
There was a discussion about putting errors were they happen before, instead of all in a single file were they become un-linked to their code. 
What's your take on this?

grumbach commented on autonomi/src/client/mod.rs:
```diff
@@ -52,33 +60,70 @@ const CLIENT_EVENT_CHANNEL_SIZE: usize = 100;
     pub(crate) network: Network,
     pub(crate) client_event_sender: Arc<Option<mpsc::Sender<ClientEvent>>>,
     pub(crate) evm_network: EvmNetwork,
+    pub(crate) mode: ClientMode,
```
What's the reason behind mode here?

dirvine replied:
The mode is for local network mode (mens) or not.

grumbach commented on autonomi/src/client/pointer.rs:
```diff
@@ -35,20 +39,17 @@ pub enum PointerError {
                 Ok(pointer)
             }
-            None => Err(PointerError::Corrupt),
+            None => Err(PointerError::FailedVerification),
```
Does `FailedVerification` cover all possible errors behind `get_local_record`?
I think having at least a String with some information on the actual underlying error would be extremely valuable

grumbach commented on autonomi/src/client/utils.rs:
```diff
@@ -17,82 +19,71 @@ use bytes::Bytes;
+        let get_chunks = move |xor_names: &[XorName]| -> Result<Vec<Bytes>, SelfEncryptionError> {
+            let mut chunks = Vec::with_capacity(xor_names.len());
+            for xor_name in xor_names {
+                match futures::executor::block_on(client.chunk_get(*xor_name)) {
```
Why not just await here instead of `block_on`?

grumbach commented on autonomi/src/network/local.rs:
```diff
@@ -0,0 +1,173 @@
+}
+
+impl LocalNode {
+    pub async fn start() -> Result<Self> {
+        Self::start_internal(true).await
+    }
+
+    pub async fn start_secondary() -> Result<Self> {
+        Self::start_internal(false).await
+    }
```
I'm not sure what `start` and `start_secondary` involves here, could this be precised?

dirvine committed:
fix: handle local feature flag in tests and fix record store tests

grumbach commented on autonomi/tests/register.rs:
```diff
@@ -22,7 +22,7 @@ use tokio::time::sleep;
     let _log_appender_guard = LogBuilder::init_single_threaded_tokio_test("register", false);
 
-    let client = Client::init_local().await?;
+    let client = Client::init_local(true).await?;
```
It's not 100% clear what the `true` here means, if an option on/off is needed maybe another constructor or the following pattern:
```
Client::init_local().with_option_AAA().await?;
```

grumbach reviewed (commented):
Lots of great additions! Having docs in here is a real game changer!
Couple of notes and questions but overall excellent work!

dirvine committed:
refactor: remove encrypt-records feature flag and make encryption mandatory

- Remove encrypt-records feature flag from ant-networking and ant-node\n- Make record encryption mandatory in record_store.rs\n- Update open-metrics feature to include prometheus-client\n- Remove redundant test steps in CI workflow

dirvine committed:
refactor: remove fs and encrypted_records feature flags - Remove fs feature flag as filesystem operations are now core functionality - Remove encrypted_records feature flag as encryption is now always enabled - Update tests and workflows to remove feature flag references - Clean up related code in client files and test modules

dirvine commented:
closing as it's too big and the local flag is leading to confusion here
